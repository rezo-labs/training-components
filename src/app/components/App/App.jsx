import React from 'react';
import styled, { keyframes } from 'styled-components';

const Wrapper = styled.div`
    padding: 150px 0;
    background: linear-gradient(45deg, rgba(170,0,255,1) 0%, rgba(25,117,210,1) 50%, rgba(32,148,243,1) 100%);
`;

const fadeIn = keyframes`
    from {
        opacity: 0;
    }

    to {
        opacity: 1;
    }
`;

const H1 = styled.h1`
    color: white;
    font-size: 2rem;
    font-weight: bold;
    margin: 0;
    padding: 10px;
    text-align: center;
    animation: ${fadeIn} 1s ease-in-out;
`;

const H2 = styled(H1)`
    font-size: 1rem;
`;

export default function App() {
    return (
        <Wrapper>
            <H1>Hello World!</H1>
            <H2 as="h2">Let&lsquo;s build something great</H2>
        </Wrapper>
    );
}
