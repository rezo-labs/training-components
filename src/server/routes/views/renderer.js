import React from 'react';
import { renderToString } from 'react-dom/server';
import { StaticRouter } from 'react-router-dom';
import { ChunkExtractor } from '@loadable/server';
import { ServerStyleSheet, StyleSheetManager } from 'styled-components';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { Helmet } from 'react-helmet';
import chalk from 'chalk';
import path from 'path';
import express from 'express'; /* eslint no-unused-vars: "off" */

import {
    state,
} from '../../../app/redux/reducers';
import App from '../../../app/components/App';
import Head from '../../../app/components/Head';
import GlobalStyle from '../../../app/components/GlobalStyle';

import compilationStats from '../../../../dist/compilation-stats.json';
import stats from '../../../../dist/loadable-stats.json';

const extractor = new ChunkExtractor({ stats });

/**
 * Metadata object for rendering head tag.
 * @typedef {Object} Metadata
 * @property {string} title - The title of the page
 * @property {string} description - The description of the page
 * @property {string} keywords - The keywords of the page
 * @property {string} canonical - The canonical of the page
 */

/**
 * Rendering function WITHOUT SSR.
 * @param {express.Request} req - The request object.
 * @param {express.Response} res - The response object.
 * @param {String} ejs - The name of the template file to be rendered.
 * @param {Metadata} metadata - The metadata object.
 * @param {Object} [initState={}] - The initial state of Redux store.
 * @param {Number} [status=200] - The status code to response.
 * @param {Object} [others={}] - Other metadata information to render the template
 *  such as description, keywords, etc.
 */
const renderWithoutSSR = (req, res, ejs, metadata, initState = {}, status = 200, others = {}) => {
    const staticContext = { statusCode: status };
    const store = createStore(state, {
        ...initState,
        metadata,
    });
    const data = {
        html: '',
        scriptTags: `<script src="/build/main.bundle${process.env.NODE_ENV === 'production'
            ? `.${compilationStats.hash}` : ''}.js"></script>`,
        preloadedState: store.getState(),
        // TODO: don't let these empty
        styleTags: '',
        helmet: {
            htmlAttributes: '',
            title: '',
            meta: '',
            link: '',
            bodyAttributes: '',
        },
        others,
    };
    res.status(staticContext.statusCode);
    res.render(ejs, { data });
};

/**
 * Rendering function WITH SSR.
 * @param {express.Request} req - The request object.
 * @param {express.Response} res - The response object.
 * @param {String} ejs - The name of the template file to be rendered.
 * @param {Metadata} metadata - The metadata object.
 * @param {Object} initState - The initial state of Redux store.
 * @param {Number} [status=200] - The status code to response.
 * @param {Object} [others={}] - Other metadata information to render the template
 *  such as description, keywords, etc.
 */
const renderWithSSR = (req, res, ejs, metadata, initState, status = 200, others = {}) => {
    const staticContext = { statusCode: status };
    const store = createStore(state, {
        ...initState,
        metadata,
    });
    const sheet = new ServerStyleSheet();

    const jsx = extractor.collectChunks(
        <StyleSheetManager sheet={sheet.instance}>
            <StaticRouter location={req.url} context={staticContext}>
                <Provider store={store}>
                    <>
                        <Head />
                        <GlobalStyle />
                        <App />
                    </>
                </Provider>
            </StaticRouter>
        </StyleSheetManager>,
    );

    let html = '';
    let styleTags = '';
    try {
        html = renderToString(jsx);
        styleTags = sheet.getStyleTags();
    }
    catch (error) {
        // handle error
        console.error(error);
    }
    finally {
        sheet.seal();
    }

    const helmet = Helmet.renderStatic();

    const scriptTags = extractor.getScriptTags();

    const data = {
        html,
        styleTags,
        scriptTags,
        preloadedState: store.getState(),
        helmet,
        others,
    };

    res.status(staticContext.statusCode);
    res.render(ejs, { data });
};

let render = renderWithoutSSR; // eslint-disable-line import/no-mutable-exports
if (process.env.SSR) {
    console.log(chalk.blue.bold('SERVER-SIDE RENDERING ON!'));
    render = renderWithSSR;
}

export default render;
