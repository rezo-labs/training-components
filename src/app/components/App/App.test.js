
import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import App from './App';

describe('App', () => {
    const tree = renderer.create(<App />).toJSON();

    it('should render as expected', () => {
        expect(tree).toMatchSnapshot();
    });
});
