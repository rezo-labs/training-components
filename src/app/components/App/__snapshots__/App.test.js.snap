// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render as expected 1`] = `
.c0 {
  padding: 150px 0;
  background: linear-gradient(45deg,rgba(170,0,255,1) 0%,rgba(25,117,210,1) 50%,rgba(32,148,243,1) 100%);
}

.c1 {
  color: white;
  font-size: 2rem;
  font-weight: bold;
  margin: 0;
  padding: 10px;
  text-align: center;
  -webkit-animation: eMLfYp 1s ease-in-out;
  animation: eMLfYp 1s ease-in-out;
}

.c2 {
  color: white;
  font-size: 2rem;
  font-weight: bold;
  margin: 0;
  padding: 10px;
  text-align: center;
  -webkit-animation: eMLfYp 1s ease-in-out;
  animation: eMLfYp 1s ease-in-out;
  font-size: 1rem;
}

<div
  className="c0"
>
  <h1
    className="c1"
  >
    Hello World!
  </h1>
  <h2
    className="c2"
  >
    Letâ€˜s build something great
  </h2>
</div>
`;
